python/GafferUI/MessageWidget.py

```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+
+
+# Provides filtering based on message level. This isn't directly used for search
+# filtering as we often want search not to affect the number messages displayed.
+class _MessageTableFilterModel( QtCore.QSortFilterProxyModel ) :
+
+	def __init__( self, level = IECore.MessageHandler.Level.Info, *kw ) :
+
+		QtCore.QSortFilterProxyModel.__init__( self, *kw )
+		self.setLevel( level )
+
+	def setLevel( self, level ) :
+
+		self.__maxLevel = level
+		self.invalidateFilter()
+
+	def getLevel( self ) :
+
+		return self.__maxLevel
+
+	# Overrides for methods inherited from QSortFilterProxyModel
+	# --------------------------------------------------------
+
+	def filterAcceptsRow( self, sourceRow, sourceParent ) :
+
+		levelIndex = self.sourceModel().index( sourceRow, _MessageTableModel.Column.Level, sourceParent )
+		return self.sourceModel().data( levelIndex, _MessageTableModel.ValueRole ) <= self.__maxLevel
+
+# The base TabelModel representing the underlying message data.
+class _MessageTableModel( QtCore.QAbstractTableModel ) :
+
+	Column = IECore.Enum.create( "Level", "Context", "Message" )
+
+	# A role to allow access the underlying Message data, without any display coercion.
+	ValueRole = 100
+
+	def __init__( self, messages = None, parent = None ) :
+
+		QtCore.QAbstractTableModel.__init__( self, parent )
+
+		self.__messages = messages
+
+	def setMessages( self, messages ) :
+
+		# We make use of existing rows here rather than resetting
+		# the model as it avoids flickering where the view first
+		# scrolls to the top, and then is re-scrolled back to the
+		# bottom.
+
+		numRows = len( self.__messages ) if self.__messages else 0
+		targetNumRows = len( messages )
+
+		if targetNumRows > numRows :
+
+			self.beginInsertRows( QtCore.QModelIndex(), numRows, targetNumRows - 1 )
+			self.__messages = messages
+			self.endInsertRows()
+
+		elif targetNumRows < numRows :
+
+			self.beginRemoveRows( QtCore.QModelIndex(), targetNumRows, numRows - 1 )
+			self.__messages = messages
+			self.endRemoveRows()
+
+		else :
+
+			self.__messages = messages
+
+		if targetNumRows > 0 :
+			self.dataChanged.emit(
+				self.index( 0, 0 ),
+				self.index( targetNumRows - 1, self.columnCount() - 1 )
+			)
+
+	def getMessages( self ) :
+
+		return self.__messages
+
+	def addMessage( self, message ) :
+
+		nextIndex = self.rowCount()
+		self.beginInsertRows( QtCore.QModelIndex(), nextIndex, nextIndex )
+		self.__messages.add( message )
+		self.endInsertRows()
+
+	# Overrides for methods inherited from QAbstractTableModel
+	# --------------------------------------------------------
+
+	def rowCount( self, parent = QtCore.QModelIndex() ) :
+
+		if parent.isValid() :
+			return 0
+
+		if self.__messages is None :
+			return 0
+
+		return len( self.__messages )
+
+	def columnCount( self, parent = QtCore.QModelIndex() ) :
+
+		if parent.isValid() :
+			return 0
+
+		return len( _MessageTableModel.Column.values() )
```

Comment:
This is showing up as a bit of a hotspot in VTune. Seems like it makes and throws away a tuple every time. If we stick with a Python implementation, might be worth using a constant here instead.

Reply:
https://github.com/GafferHQ/gaffer/pull/3790/commits/3749bef030b027cf04a097b9e7f04f778a0e16c2